TypeScript/HW_12/task 12.2.1.ts

```diff
@@ -0,0 +1,25 @@
+// Task 2
+// 1. Напишите функцию, реализующую методы массивов map. Функции принимают на вход массив и колбэк. Используйте дженерик типы. 
+//    Затипизировать надо саму функцию и коллбэк.
+//    Создать реализацию функции map, принимающую массив чисел 1-5, возвращающую новый массив, 
+//    где каждый каждый элемент - это элемент исходного массива умноженный на его индекс
+//    Пример:
+//    map([1,2,3,4,5], callback) => [0,2,6,12,20]
+
+type MapCallback<T> = (value: T, index: number, array: T[]) => T;
+
+function map<T>(array: T[], callback: MapCallback<T>): T[] {
+  let result: T[] = [];
+  for (let i = 0; i < array.length; i++) {
+    result.push(callback(array[i], i, array));
+  }
+  return result;
+};
```

Comment:
For `map` function we need two types in generic because method is used to transform an array of one type into an array of another type

```suggestion
function map<T, U>(array: T[], callback: MapCallback<T>): U[] {
  let result: U[] = [];
  for (let i = 0; i < array.length; i++) {
    result.push(callback(array[i], i, array));
  }
  return result;
};
```

Reply:
Thanks, I figured it out, fixed